### ISSUE TITLE
mine doesn't work with multiple masters

### ISSUE THREAD

**Issue by ozgurakan** [saltstack/salt#7697]
When I have

```
master:
- master1.domain.com
- master2.domain.com
```

or

```
master:
- master1.domain.com
```

mine code in Jinja templates or SLS formulas stop working.

When I have

```
master: master1.domain.com
```
mine works fine.

Sample code in a jinja template file that get affected;

```
{%-                     for key, value in salt['mine.get']('environment_id:' + grains['environment_id'], 'grains.items', expr_form='grain').items() %}
{%-                         if role in value['roles'] %}
-A INPUT -s {{ key }} -j {{ role|upper }}
{%-                         endif %}
{%-                     endfor %}
```

```
master:t# salt-call --versions-report (ubuntu)
           Salt: 0.16.4
         Python: 2.7.4 (default, Apr 19 2013, 18:28:01)
         Jinja2: 2.6
       M2Crypto: 0.21.1
 msgpack-python: 0.2.0
   msgpack-pure: Not Installed
       pycrypto: 2.6
         PyYAML: 3.10
          PyZMQ: 13.0.0
            ZMQ: 3.2.2
```

```
minion:# salt-call --versions-report (centos)
           Salt: 0.16.4
         Python: 2.6.6 (r266:84292, Jul 10 2013, 22:48:45)
         Jinja2: unknown
       M2Crypto: 0.20.2
 msgpack-python: 0.1.13
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
         PyYAML: 3.10
          PyZMQ: 2.2.0.1
            ZMQ: 3.2.4
```

thanks,
Oz

**Comment by BrianGallew** [comment]
This is still kind of an important problem.  And by "kind of" I mean "mine is non-functional in any non-trivial environment".

**Comment by rgardam** [comment]
Any updates on this. 

It would be great to be able to read and write mine data to a select number of master nodes. The idea is that the mine could be separate from the master nodes.

**Comment by onorua** [comment]
Any updates? Does it mean that I must "switch off" my second master to get mine working? I've spent almost 3 days trying to figure out what is wrong until I found this bug... maybe you could point this out in documentation at least so people like I could know the limitation of the platform? But better get it working :)

**Comment by onorua** [comment]
I was thinking a bit, looking at the code. I discovered that in case of multimaster, you  have only one master connected and the rest are keep just in case of failure, and if nothing happened to primary one - you generally don't care about the rest. Which is good approach to save open connections and speed of the processing. 
So in order to propagate mine data from minions to masters, we have two options:
 1. Send mine data to each and every master
 2. Have synchronisation mechanism between masters. 

I prefer option 2 much more because:
 - it keeps amount of client connections small enough and with it random function does make sense. 
 - it keeps speed of mine synchronisation the same or almost the same. 
 - in case one of the masters was unavailable, it will be synchronized with the rest during startup, preventing corrupted data in case of multimaster with random function enabled in a moment between mine updates. 
 - we don't overcomplicate minions, and it will be possible to upgrade masters keeping minions which can't be updated due to some limitations ( I have couple of such systems, where each update must be agreed with 2-3 parties and it takes like a week to get it approved). 

What can be done in this case:
  1. Create synchronous replication mechanism between masters
 * With replication, we keep things simple, but we don't solve problem when one of the masters is down for some time.
  2. Some clustered filesystem to keep common cache. 
 * With clustered filesystem we solve all the problems. 

I can be wrong, because I did this conclusions based on glance review of the  code.  I would like to help to solve this issue, but I would like to hear from you guys, before I go to reinvent the wheel.

**Comment by rgardam** [comment]
I would suggest using a technology that already has sharding and replication built in rather than trying to reinvent the wheel. 

Shared FS's can be a pretty evil. 

Something like Elasticsearch comes to mind.

**Comment by onorua** [comment]
@rgardam seriously? Elasticsearch to store almost static data which must be just synced together? I believe it's overkill...

**Comment by CrackerJackMack** [comment]
@rgardam Elastic search can be used to store billions of static documents for an infinite amount of time to perform later data analysis.  It's consistencies in regards to a mine data target, its a [good distributed document store](https://aphyr.com/posts/323-call-me-maybe-elasticsearch-1-5-0).  Before we get into the "this distributed data store is better" debate there might be different use cases where last-write wins is okay and in other cases where it is intolerable.  This [blog series](https://www.google.com/search?q=aphyr+call+me+maybe) helps to enlighten the end-users of NoSQL/distributed document store databases/[CAP theorem](https://en.wikipedia.org/wiki/CAP_theorem) [hype](http://arxiv.org/abs/1509.05393) databases.  

Since mine data, grain data, and all of the above are just document stores (or better summarized as multiple KV pairs namespaced by the minion_id) it would make sense that all of these have a pluggable backend.  Most of the framework in there in [job caches and returners](https://docs.saltstack.com/en/latest/ref/returners/all/index.html#all-salt-returners). 

IMO it makes the most sense that this problem should be moved downstream away from salt to manage directly and into subject matter experts to handle these types of scale out issues.

Truth be told, not everyone will experience these large scale out issues and thus salt-master only needs to facilitate a method in which scaling can occur rather than solve the issue by re-inventing the wheel in regards to caches.  Again, the returners and job cache dealt with this by moving the data storage problem downstream away from the salt-master and into plugins.  (I'm aware you can still send salt returners to the master and the master will store the data in the backend, but the code is in plugins and configurable instead of being inherently entangled in the salt-master codebase).

All that being said, elasticsearch has really good ways to fetch a lot of documents based on the needs of the compound matcher.  I am ignorant to etcd/consoul/riak/couchdb/cassandra/mongodb lookup mechanisms and how they would accomplish such a task so I am *not* down playing their capabilities and trumpeting ES as being the only viable contender.

**Comment by rgardam** [comment]
@onorua - I can understand where you're coming from. I wasn't meaning use elasticsearch, I was meaning use some key/value store that already handles CAP for you. Shared file systems just make me shudder a little bit. The only reason i suggested ES was due to my experience with it.

@CrackerJackMack - I agree. These things should be pluggable, but salt also needs to be able to handle one or many (insert internal salt service). It seems like some services work with multiple masters and some behave in an unexpected fashion. Some consistency would be nice.

**Comment by CrackerJackMack** [comment]
Is this now addressed by #27446 as well ? Or maybe future work that can be ported to the masters ?

**Comment by basepi** [comment]
Yep, I was actually about to comment, I think I'm going to close this as a duplicate of https://github.com/saltstack/salt/issues/27446, which will solve this issue.
